Improve the eSPI emulators

barnas-michal opened the pull request:
These commits refactor the eSPI emulators and implement the missing logic for the LPC-over-eSPI requests and also the built-in 8042 keyboard controller.

barnas-michal committed:
eSPI: move espi_utils.h to include directory

Move the espi_utils.h header file to include directory to allow
it to be included from the tests.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
eSPI: improve the eSPI bus and host emulators

Refactor the eSPI emulator to better match the MCU interface.
Eliminate the chip-selects which doesn't suit the peripheral-mode
of eSPI API. Make it more like a point-to-point bus.
Implement the LPC-over-eSPI requests and 8042 keyboard controller
logic.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

albertofloyd commented:
LGTM could you indicate where eSPI host emulator is currently exercised? Any available sample code

semihalf-barnas-michal commented:
> LGTM could you indicate where eSPI host emulator is currently exercised? Any available sample code

There is code that uses the emulators: https://chromium-review.googlesource.com/c/chromiumos/platform/ec/+/3613913
Files that implement the tests logic are in path: zephyr/test/drivers/src/

aaronemassey commented:
> > LGTM could you indicate where eSPI host emulator is currently exercised? Any available sample code
> 
> There is code that uses the emulators: https://chromium-review.googlesource.com/c/chromiumos/platform/ec/+/3613913 Files that implement the tests logic are in path: zephyr/test/drivers/src/

For Zephyr Project code that'd be:

*  [emul_espi_host.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/emul/espi/emul_espi_host.c )
* [espi drivers test](https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/espi/)

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -19,168 +19,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+	struct espi_emul_data *data = dev->data;
+	struct espi_emul *emul = data->emul;
+
+	if (!(data->channels_enabled & ESPI_CHANNEL_VWIRE) || !emul) {
```
Use explicit `== 0` and `== NULL`

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -19,168 +19,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
 	if (!emul) {
-		LOG_INF("espi_emul not found");
 		return -EIO;
 	}
 
+	__ASSERT_NO_MSG(emul);
 	__ASSERT_NO_MSG(emul->api);
-	__ASSERT_NO_MSG(emul->api->get_vw);
-	api = emul->api;
+	__ASSERT_NO_MSG(emul->api->manage_callback);
```
Use explicit comparisons to `NULL`

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -203,28 +521,36 @@ int espi_emul_register(const struct device *dev, const char *name,
 		.config = espi_emul_config,
 		.get_channel_status = espi_emul_get_channel_status,
-		.read_lpc_request = espi_emul_read_lpc_request,
-		.write_lpc_request = espi_emul_write_lpc_request,
+		.read_request = espi_emul_read_request,
+		.write_request = espi_emul_write_request,
+		.read_lpc_request = espi_emul_lpc_read_request,
+		.write_lpc_request = espi_emul_lpc_write_request,
 		.send_vwire = espi_emul_send_vwire,
 		.receive_vwire = espi_emul_receive_vwire,
-		.manage_callback = espi_emul_manage_callback
+		.send_oob = espi_emul_send_oob,
+		.receive_oob = espi_emul_receive_oob,
+		.flash_read = espi_emul_flash_read,
+		.flash_write = espi_emul_flash_write,
+		.flash_erase = espi_emul_flash_erase,
+		.manage_callback = espi_emul_manage_callback,
```
It looks like some of there don't have the `static` keyword, do they need to have external linkage?

semihalf-barnas-michal replied:
Only some of them needs to be visible, fixed

yperess commented on include/zephyr/drivers/espi/espi_utils.h:
```diff
@@ -5,7 +5,7 @@
  */
 
 /**
- * @file Header with definitions for eSPI drivers callback functions
+ * Header file with definitions for eSPI drivers callback functions
  */
 
 #ifndef ZEPHYR_DRIVERS_ESPI_UTILS_H_
```
Why move this file? This doesn't seem to be needed as a publicly facing API

semihalf-barnas-michal replied:
Public API doesn't require it, but the emul subsys do (espi_emul_host). I moved it to be able to include using directive
`#include <drivers/espi/espi_utils.h>`
instead of
`#include "../../../drivers/espi/espi_utils.h"`
since the root of source code is not in the include paths

yperess commented on include/zephyr/drivers/espi_emul.h:
```diff
@@ -116,20 +53,65 @@ struct espi_emul {
+	/* Virtual wire signal */
+	enum espi_vwire_signal sig;
+	/* The level(state) of the virtual wire */
+	uint8_t level;
```
Are there units to this?

semihalf-barnas-michal replied:
Done

yperess commented on include/zephyr/drivers/espi_emul.h:
```diff
@@ -116,20 +53,65 @@ struct espi_emul {
+	/* The level(state) of the virtual wire */
+	uint8_t level;
+	/* The direction of the virtual wire. Possible values are
+	 * ESPI_MASTER_TO_SLAVE or ESPI_SLAVE_TO_MASTER
```
Inclusive language, please rephrase to `controller`/`target`

semihalf-barnas-michal replied:
It's using the defines from `espi.h`, do we want to change them into enums?

alevkoy replied:
I would approve of converting the constants in `espi.h` to enums, but that's probably out of scope here.

semihalf-barnas-michal replied:
After some refactoring, they were redundant, so I've just removed this field and usage of these defines.

yperess commented on subsys/emul/espi/espi_emul_host.c:
```diff
@@ -0,0 +1,388 @@
+#include <zephyr/drivers/espi_emul.h>
+
+/** Declare the default state of virtual wires */
+const static struct espi_emul_vw_data vw_state_default[] = {
+	{ ESPI_VWIRE_SIGNAL_SLP_S3,		0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_SLP_S4,		0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_SLP_S5,		0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_OOB_RST_WARN,	0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_PLTRST,		0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_SUS_STAT,		0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_NMIOUT,		0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_SMIOUT,		0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_HOST_RST_WARN,	0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_SLP_A,		0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_SUS_PWRDN_ACK,	0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_SUS_WARN,		0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_SLP_WLAN,		0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_SLP_LAN,		0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_HOST_C10,		0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_DNX_WARN,		0, ESPI_MASTER_TO_SLAVE },
+};
```
Same here as `espi_emul.c`, please use the dot notation for the specific fields

semihalf-barnas-michal replied:
Done

yperess commented on subsys/emul/espi/espi_emul_host.c:
```diff
@@ -0,0 +1,388 @@
+					  uint32_t *value)
+{
+#ifdef CONFIG_ESPI_PERIPHERAL_CUSTOM_OPCODE
+	struct espi_host_emul_data *data =
+		CONTAINER_OF(dev, struct espi_host_emul_data, emul);
```
@aaronemassey Does this affect your emulator API change?

aaronemassey replied:
It will, but it'll only be minor changes to function signatures on my part.

yperess commented on subsys/emul/espi/espi_emul_host.c:
```diff
@@ -0,0 +1,388 @@
+	return 0;
+}
+
+static struct espi_emul_device_api espi_emul_host_api = {
+	.config = host_espi_api_config,
+	.get_channel_status = host_espi_api_get_channel_status,
+	.read_request = host_espi_api_read_request,
+	.write_request = host_espi_api_write_request,
+	.read_lpc_request = host_espi_api_lpc_read_request,
+	.write_lpc_request = host_espi_api_lpc_write_request,
+	.send_vwire = host_espi_api_send_vwire,
+	.receive_vwire = host_espi_api_receive_vwire,
+	.internal_set_vwire = host_espi_api_internal_set_vwire,
+	.send_oob = host_espi_api_send_oob,
+	.receive_oob = host_espi_api_receive_oob,
+	.flash_read = host_espi_api_flash_read,
+	.flash_write = host_espi_api_flash_write,
+	.flash_erase = host_espi_api_flash_erase,
+	.manage_callback = host_espi_api_manage_callback,
+	.raise_event = host_espi_api_raise_event,
+};
```
Some of there don't have `static`, is that on purpose?

semihalf-barnas-michal replied:
No, fixed

yperess reviewed (changes requested):
This is really good, mostly nits

alevkoy commented on include/zephyr/drivers/espi_emul.h:
```diff
@@ -144,42 +126,61 @@ int espi_emul_register(const struct device *dev, const char *name,
+ *           the direction of vwire is invalid
  */
-int emul_espi_host_port80_write(const struct device *espi_dev, uint32_t data);
+int espi_emul_internal_set_vwire(const struct device *dev,
```
What does "internal" mean here? Should this be part of the API for this module?

semihalf-barnas-michal replied:
It means that internal in the emulation sense. Let's think about the host, it emulates that it's internal value of virtual wire has changed, and it has to send this information through the eSPI bus to "us" (EC).

alevkoy replied:
So you are distinguishing between the current value of the virtual wire as reported to the EC (`send_vwire`?) and the value that the emulated host has attempted to set (`internal_set_vwire`)?

semihalf-barnas-michal replied:
Not exactly. The send_vwire function sends the virtual wire value from the EC to the other side, e.g. host. It's used to inform the host about change in wire connected to the EC. The receive_vwire is used by EC to read the value of wire connected to the host. So this internal_set_vwire is like a send_vwire function but from the host side to the EC. If the AP wants to go to sleep mode, it sends the vwire slp signals through the eSPI to the EC. This internal function emulates this.

alevkoy replied:
Would `host_set_vwire` make sense as a name? To me at least, that would be clearer.

semihalf-barnas-michal replied:
It would make much sense in this case. The question is, if our "secondary" (host) side will always be a host? Maybe it is a good idea to change that, even if there will be other emulators in the future,

alevkoy replied:
In real life, do the EC and on-chip keyboard controller communicate over the eSPI bus or using some other channel inside the EC chip? If there was an off-chip keyboard controller, would it talk to the EC at all?

If the EC and keyboard controller wouldn't actually communicate over the bus, then maybe there should be two layers of API or an otherwise slightly different structure here. (I have been kind of wondering why the keyboard controller emulation was part of the eSPI emulation module, so thank you for elaborating.)

alevkoy replied:
IIUC, eSPI is specifically intended to be used in the same context as LPC, which is intended for CPUs to communicate with peripherals on motherboards. So I think unless Zephyr is running on the CPU, the "other" side of this emulator will always be the CPU. To the extent that we use the API to communicate with the keyboard controller, it's outside the bounds of eSPI as a technology, and we can probably expect that we won't need to make similar accommodations for anything else.

I would prefer `host_set_vwire` as a name for this function and API function pointer. "cpu" or "ap" would also be reasonable choices. I don't think "secondary" or "internal" convey clearly what this function does. Same idea for the comments that use "secondary."

alevkoy commented on include/zephyr/drivers/espi_emul.h:
```diff
@@ -144,42 +126,61 @@ int espi_emul_register(const struct device *dev, const char *name,
  */
-uintptr_t emul_espi_host_get_acpi_shm(const struct device *espi_dev);
-#endif
+int espi_emul_manage_host_callback(const struct device *dev,
```
"Manage" makes it a bit non-obvious what this does. IMO, `espi_emul_enable_host_callback` would be clearer.

semihalf-barnas-michal replied:
It uses the zephyr's generic name of `manage_callback`. All APIs are using this naming for this scheme. Only change here is the `host` part.

alevkoy commented on include/zephyr/drivers/espi_emul.h:
```diff
@@ -144,42 +126,61 @@ int espi_emul_register(const struct device *dev, const char *name,
+ * @param ev structure describing event about which the host will be informed
+ * @return always 0
+ */
+int espi_emul_raise_event(const struct device *dev, struct espi_event ev);
```
Is this something that the emulator would be expected to call spontaneously, or will it be called from outside the emulator? Should it be part of the API?

semihalf-barnas-michal replied:
This function is used only to trigger the callbacks for tests. In normal way, these callbacks are triggered by the interrupts from the MCU, but in tests they have to be called manually from the tests.
Its usage designed to be used by e.g. host emulator, which calls it in response to internal change of vwires, simulating the process of sending the vwire change through eSPI bus to the MCU.

alevkoy replied:
It might be helpful to distinguish in the API, either with separate structs or just comments, which functions are expected to be called from the EC (or whatever the generic name for the UUT would be in this context) and which are are expected to be called from the (emulated) host.

alevkoy commented on subsys/emul/espi/espi_emul_host.c:
```diff
@@ -0,0 +1,388 @@
+
+#ifdef CONFIG_ESPI_PERIPHERAL_CUSTOM_OPCODE
+	for (a = 0; a < CONFIG_ESPI_EMUL_HOST_CMD_SHM_REGION_SIZE; a++) {
+		data->shm_host_cmd[a] = 0;
```
Could this just be something like `memset(data->shm_host_cmd, sizeof(data->shm_host_cmd), 0);`?

semihalf-barnas-michal replied:
Sure, done

alevkoy commented on include/zephyr/drivers/espi_emul.h:
```diff
@@ -116,20 +53,65 @@ struct espi_emul {
+	uint8_t dir;
+};
+
+struct espi_emul_device_api {
```
Can you comment the expected behavior, parameters, return value, etc. of each of these API functions? I see comments on at least some of the implementations. I'd prefer to have comments here and no comments there, since the interface is here.

semihalf-barnas-michal replied:
These functions are the same as for normal eSPI device, they are only required because the first parameter id espi_emul instead of device. Do we want to duplicate the comments?

alevkoy replied:
I guess not, but maybe include a reference to `struct espi_driver_api` here to point people like me in the right direction.

semihalf-barnas-michal replied:
I've updated the documentation, please take a look if it is better now

alevkoy replied:
The doc changes looks good.

As a nit, I think it would be helpful to group the members of this struct by who is expected to call them:

1. The EC. These would be the functions that have a direct analog in `espi_driver_api`, in the order they appear there.
2. The emulated bus. I think this would be `io_read` and `io_write`, but I'm not exactly clear on where those will be called from.
3. The emulated host. I think these are `internal_set_vwire` and `raise_event`.

semihalf-barnas-michal replied:
The io_* and set_vwire and raise_event are all for the host emulator. With the updated documentation, this order looks nice, done.

alevkoy commented on drivers/espi/espi_emul.c:
```diff
@@ -19,168 +19,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+#include <zephyr/drivers/espi/espi_utils.h>
+
+/** Declare the default state of virtual wires */
+const static struct espi_emul_vw_data vw_state_default[] = {
```
IIRC, ESPI defines a specific set of named vwires, and a system can use them or not use them. (I don't think this is all of them?) I have a doc to emulate VBUS in a way that is devicetree configurable. But vwires (ESPI) are unrelated to VBUS (USB).

If we were to configure how we use vwires based on devicetree, I would imagine it would have something to do with our GPIO config, like if we have a real SLP_S3, we wouldn't use the SLP_S3 vwire. But that's out of scope here.

semihalf-barnas-michal replied:
Yes, the eSPI specification defines exactly what virtual-wires exist. There is no way to define custom ones.
It is due to the protocol, since the information about level change on vwire can't be sent separately, the messages for one vwire includes also some other vwires, but marked as "invalid data". The specification describes exactly which vwires are sent with which other.

alevkoy commented on drivers/espi/espi_emul.c:
```diff
@@ -21,166 +21,470 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+
 /** Working data for the controller */
 struct espi_emul_data {
 	/* List of struct espi_emul associated with the device */
```
Should this comment change? Or is this structure still a list?

semihalf-barnas-michal replied:
Forgot this, done

alevkoy commented on drivers/espi/espi_emul.c:
```diff
@@ -19,168 +19,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+	/** Virtual wires */
+	struct espi_emul_vw_data vw_state[NUMBER_OF_VWIRES];
+	/** Channel configuration */
+	uint8_t channels_enabled;
```
If this is to be used as a bitset, can you comment that here?

semihalf-barnas-michal replied:
Changed to enum, which is assigned to this field

semihalf-barnas-michal commented on include/zephyr/drivers/espi/espi_utils.h:
```diff
@@ -5,7 +5,7 @@
  */
 
 /**
- * @file Header with definitions for eSPI drivers callback functions
```
The doxygen complained that the description is invalid. It wasn't checked when file wasn't in include directory

semihalf-barnas-michal commented on drivers/espi/espi_emul.c:
```diff
@@ -19,168 +19,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+
+/** Declare the default state of virtual wires */
+const static struct espi_emul_vw_data vw_state_default[] = {
+	{ ESPI_VWIRE_SIGNAL_OOB_RST_ACK,   0, ESPI_SLAVE_TO_MASTER },
+	{ ESPI_VWIRE_SIGNAL_WAKE,          0, ESPI_SLAVE_TO_MASTER },
+	{ ESPI_VWIRE_SIGNAL_PME,           0, ESPI_SLAVE_TO_MASTER },
+	{ ESPI_VWIRE_SIGNAL_SLV_BOOT_DONE, 0, ESPI_SLAVE_TO_MASTER },
+	{ ESPI_VWIRE_SIGNAL_ERR_FATAL,     0, ESPI_SLAVE_TO_MASTER },
+	{ ESPI_VWIRE_SIGNAL_ERR_NON_FATAL, 0, ESPI_SLAVE_TO_MASTER },
+	{ ESPI_VWIRE_SIGNAL_SLV_BOOT_STS,  0, ESPI_SLAVE_TO_MASTER },
+	{ ESPI_VWIRE_SIGNAL_SCI,           0, ESPI_SLAVE_TO_MASTER },
+	{ ESPI_VWIRE_SIGNAL_SMI,           0, ESPI_SLAVE_TO_MASTER },
+	{ ESPI_VWIRE_SIGNAL_RST_CPU_INIT,  0, ESPI_SLAVE_TO_MASTER },
+	{ ESPI_VWIRE_SIGNAL_HOST_RST_ACK,  0, ESPI_SLAVE_TO_MASTER },
+	{ ESPI_VWIRE_SIGNAL_SUS_ACK,       0, ESPI_SLAVE_TO_MASTER },
+	{ ESPI_VWIRE_SIGNAL_DNX_ACK,       0, ESPI_SLAVE_TO_MASTER },
```
Done

semihalf-barnas-michal commented on drivers/espi/espi_emul.c:
```diff
@@ -19,168 +19,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+	uint8_t kbc8042_dbbin;
+	uint8_t kbc8042_dbbout;
+	uint8_t kbc8042_status;
+	uint8_t kbc8042_irq_enabled;
```
Done

semihalf-barnas-michal commented on drivers/espi/espi_emul.c:
```diff
@@ -19,168 +19,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
 	struct espi_emul_data *data = dev->data;
 
-	return (data->cfg.channel_caps & ch);
+	if (data->channels_enabled & ch) {
+		return 0;
+	}
+
+	return -EIO;
```
Done

semihalf-barnas-michal commented on drivers/espi/espi_emul.c:
```diff
@@ -19,168 +19,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+	struct espi_emul_data *data = dev->data;
+	struct espi_emul *emul = data->emul;
+
+	if (!(data->channels_enabled & ESPI_CHANNEL_PERIPHERAL)) {
```
Done

semihalf-barnas-michal commented on drivers/espi/espi_emul.c:
```diff
@@ -19,168 +19,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+	switch (op) {
+#ifdef CONFIG_ESPI_EMUL_KBC8042
+	case E8042_OBF_HAS_CHAR:
+		*value = !!(data->kbc8042_status & KBC8042_STATUS_OBF);
```
Done

semihalf-barnas-michal commented on drivers/espi/espi_emul.c:
```diff
@@ -19,168 +19,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+		*value = !!(data->kbc8042_status & KBC8042_STATUS_OBF);
+		break;
+	case E8042_IBF_HAS_CHAR:
+		*value = !!(data->kbc8042_status & KBC8042_STATUS_IBF);
```
Hm, are you sure that it will raise the error? I've done a little test with gcc 9.4.0 and doesn't fail to compile:
```
        uint8_t x = 5;
        int y = 9;
        uint32_t v;
        uint32_t* ptr = &v;

        *ptr = !!(x & y);
        printf("V: %d", v);
```

Also it looks like the "!!(x&y)" is used often in the zephyr drivers, even in ITE and NPCX espi drivers.

semihalf-barnas-michal commented on drivers/espi/espi_emul.c:
```diff
@@ -19,168 +19,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+	struct espi_emul_data *data = dev->data;
+	struct espi_emul *emul = data->emul;
+
+	if (!(data->channels_enabled & ESPI_CHANNEL_PERIPHERAL)) {
```
Done

semihalf-barnas-michal commented on drivers/espi/espi_emul.c:
```diff
@@ -19,168 +19,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+		break;
 	default:
-		return -EINVAL;
+		return emul->api->write_lpc_request(data->emul, op, value);
```
Done

alevkoy commented on include/zephyr/drivers/espi_emul.h:
```diff
@@ -116,20 +53,61 @@ struct espi_emul {
+		enum lpc_peripheral_opcode op, uint32_t *data);
+	int (*write_lpc_request)(const struct espi_emul *dev,
+		enum lpc_peripheral_opcode op, uint32_t *data);
+	int (*io_read)(const struct espi_emul *dev,
```
What is the analog to this in espi_driver_api?

semihalf-barnas-michal replied:
Oh, ok... These don't have an analogical functions in real espi drivers. They are emulating the IO read/write through the eSPI bus which is handled internally by the EC chips. These require description.

alevkoy reviewed (dismissed):
Some minor nits. Otherwise, this looks good.

carlescufi commented:
@aaronemassey @albertofloyd please take another look

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -19,170 +19,480 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
-	/* List of struct espi_emul associated with the device */
-	sys_slist_t emuls;
+	/* eSPI emul associated with the device */
+	struct espi_emul *emul;
```
This doesn't look right to me, it needs to be a list. You can have multiple peripherals attached to the same bus.

semihalf-barnas-michal replied:
Not exactly from our side. The host can have multiple peripherals, but our EC is peripheral-only.
eSPI API doesn't contain any chip-selecting, we just send data to the bus, not to the peripheral, and the host is signaled using interrupts that it needs to read our data.
The best example is the keyboard controller which can be embedded in EC. Using our API, we send some requests to the KBC but using the bus connected to the host (bus is our destination device). Theoretically, we are sending these data to host, but the chip drivers intercepts these requests and handle them internally (so do this emulator). We could have another eSPI bus only for KBC, but then, the KBC would need to have peripheral bus to controller EC and another peripheral bus to controller host (because host needs to read the data from KBC using the I/O through eSPI/LPC).
Looking at this, from EC (zephyr) side, we can't be connected to more than one (host) peripheral on the bus.

jeremybettis replied:
Maybe we just need to make it more explicit that this is only a host emulator, not a general purpose espi emulator then.

semihalf-barnas-michal replied:
I tried to change the descriptions and Kconfigs to be more accurate. Let me know if there is more I should do with it.

jackrosenthal commented on include/zephyr/drivers/espi_emul.h:
```diff
@@ -144,42 +154,86 @@ int espi_emul_register(const struct device *dev, const char *name,
+ *           returned by the emulator
+ */
+int espi_emul_host_manage_callback(const struct device *dev,
+	struct espi_callback *callback, bool set);
```
Formatting seems funky. Try `clang-format`.

barnas-michal committed:
espi: cleanup the code

This commit refactors the common espi code using the clang format.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

fabiobaltieri commented:
@yperess can you re-review this? (you have a pending blocking comment)

keith-zephyr commented on drivers/espi/espi_emul.c:
```diff
@@ -21,170 +21,477 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+	 * is connected. This peripheral eSPI emulator doesn't emulate whole bus, instead, it
+	 * emulates the eSPI as seen by the MCU, being peripheral connected to the host controller.
+	 */
+	struct espi_emul *emul;
```
Nit - this would be clearer as `host_emul`.

keith-zephyr commented on drivers/espi/espi_emul.c:
```diff
@@ -21,170 +21,477 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+		break;
 	default:
-		return -EINVAL;
+		__ASSERT_NO_MSG(emul != NULL);
```
The other functions registered with `struct espi_driver_api` return -EIO if the host emulator isn't provided.
Is it valid to use this peripheral emulator without a host emulator?   If the host emulator is required, you could move this check into `espi_emul_init()` or `espi_emul_api_config()`

semihalf-barnas-michal replied:
Technically it's correct, you can have an eSPI peripheral which isn't connected to any host.

keith-zephyr commented on drivers/espi/espi_emul.c:
```diff
@@ -21,170 +21,477 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
 	default:
-		return -EINVAL;
+		__ASSERT_NO_MSG(emul != NULL);
+		__ASSERT_NO_MSG(emul->api != NULL);
```
Move this check into `espi_emul_register()`

fabiobaltieri commented:
@yperess can you review this?

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -17,174 +17,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+#include <zephyr/drivers/espi/espi_utils.h>
+
+/** Declare the default state of virtual wires */
+const static struct espi_emul_vw_data vw_state_default[] = {
```
nit: coding style should have `static` then `const` I believe

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -17,174 +17,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
 {
-	struct espi_emul_data *data = dev->data;
-	sys_snode_t *node;
+	unsigned int i;
```
nit: Zephyr style allows/encourages defining `i` inside the `for`

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -17,174 +17,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
-		emul = CONTAINER_OF(node, struct espi_emul, node);
-		if (emul->chipsel == chipsel) {
-			return emul;
+	for (idx = 0; idx < NUMBER_OF_VWIRES; idx++) {
```
nit: same here, `idx` can be inside the `for` loop

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -17,174 +17,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+
+	espi_send_callbacks(&data->callbacks, dev, ev);
+}
+#endif
```
nit: this is a long `ifdef`, please add a closing `/* CONFIG_ESPI_EMUL_KBC8042 */

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -17,174 +17,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+	/** Virtual wires states */
+	struct espi_emul_vw_data vw_state[NUMBER_OF_VWIRES];
+	/** Bit field of enabled channels on eSPI bus */
+	enum espi_channel channels_enabled;
```
This is confusing, below you do `channels_enabled & ch` to check if the channel is enabled, the enum is a bit field, should this be a `uint8_t`? Can more than one be set at a time?

semihalf-barnas-michal replied:
Hm, it sure can be a misleading. However, the espi subsystem defines the `struct espi_cfg` where you have this enum, with comment in plural form `Supported channels` and also the values of the enums are bits aligned. So yes, multiple values can be set at a time. I've checked and other drivers for chips are using this type in the same way.
E.g. microchip `if (cfg->channel_caps & ESPI_CHANNEL_PERIPHERAL)`

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -17,174 +17,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
 	struct espi_emul_data *data = dev->data;
 
-	return (data->cfg.channel_caps & ch);
+	return (data->channels_enabled & ch);
```
nit: implicit cast, please add `==` or `!=`.

Also, see comment above, `channels_enabled` seems to be an enum, we shouldn't be doing a bit operator on it.

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -17,174 +17,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
-	const struct emul_espi_device_api *api;
-	struct espi_emul *emul;
-	struct espi_emul_data *emul_data = dev->data;
+	return -EIO;
```
nit: add `ARG_UNUSED` for both `dev` and `req`

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -17,174 +17,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
 
-	ARG_UNUSED(data);
+static int espi_emul_api_write_request(const struct device *dev, struct espi_request_packet *req)
+{
```
nit: add `ARG_UNUSED` for `dev` and `req` here too

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -17,174 +17,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
-	if (!emul) {
-		LOG_ERR("espi_emul not found");
-		return -ENOTSUP;
+	if ((data->channels_enabled & ESPI_CHANNEL_PERIPHERAL) == 0) {
```
reference above, using enum with bit operator is odd, please revisit

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -17,174 +17,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+	int idx;
 
-	if (!(data->cfg.channel_caps & ESPI_CHANNEL_VWIRE)) {
+	if ((data->channels_enabled & ESPI_CHANNEL_VWIRE) == 0) {
```
reference above enum with bit operator

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -17,174 +17,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+	} else {
+		if (host_emul == NULL) {
+			return -EIO;
+		}
 
-	__ASSERT_NO_MSG(emul->api);
-	__ASSERT_NO_MSG(emul->api->set_vw);
-	api = emul->api;
+		return host_emul->api->send_vwire(host_emul, vw, level);
+	}
```
nit: it is generally more readable to have the error condition first so that the default case doesn't need to be inside an `if`. For example:

```
if (idx == -1) {
  if (host_emul == NULL) {
    return -EIO;
  }
  return host_emul->api->send_vwire(host_enum, vw, level);
}

data->vw_state[idx].level = level;
...
```

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -17,174 +17,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+	int idx;
 
-	if (!(data->cfg.channel_caps & ESPI_CHANNEL_VWIRE)) {
+	if ((data->channels_enabled & ESPI_CHANNEL_VWIRE) == 0) {
```
reference above, enum with bit operator

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -17,174 +17,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+	idx = espi_emul_host_find_index(data, vw);
+	if (idx != -1) {
+		*level = data->vw_state[idx].level;
+	} else {
```
Same here, check error condition with return inside, then the "normal" case doesn't need to be broken up

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -17,174 +17,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+
+/* Logical Channel 2 APIs */
+static int espi_emul_api_send_oob(const struct device *dev, struct espi_oob_packet *pckt)
+{
+	return -EIO;
+}
+
+static int espi_emul_api_receive_oob(const struct device *dev, struct espi_oob_packet *pckt)
+{
+	return -EIO;
+}
+
+/* Logical Channel 3 APIs */
+static int espi_emul_api_flash_read(const struct device *dev, struct espi_flash_packet *pckt)
+{
+	return -EIO;
+}
+
+static int espi_emul_api_flash_write(const struct device *dev, struct espi_flash_packet *pckt)
+{
+	return -EIO;
+}
 
-	return api->get_vw(emul->target, vw, level);
+static int espi_emul_api_flash_erase(const struct device *dev, struct espi_flash_packet *pckt)
+{
+	return -EIO;
```
nit: add `ARG_UNUSED` for the args

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -17,174 +17,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+
+	return 0;
+#else
+	return -EIO;
```
nit: add `ARG_UNUSED`

yperess commented on drivers/espi/espi_emul.c:
```diff
@@ -17,174 +17,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+
+	return 0;
+#else
+	return -EIO;
```
nit: add `ARG_UNUSED`

yperess commented on include/zephyr/drivers/espi/espi.h:
```diff
@@ -134,29 +134,29 @@ enum espi_pc_event {
-#define ESPI_PERIPHERAL_INDEX_0  0ul
-#define ESPI_PERIPHERAL_INDEX_1  1ul
-#define ESPI_PERIPHERAL_INDEX_2  2ul
+#define ESPI_PERIPHERAL_INDEX_0 0ul
+#define ESPI_PERIPHERAL_INDEX_1 1ul
+#define ESPI_PERIPHERAL_INDEX_2 2ul
```
nit: for these and below they should really use fixed width constants `ul` can mean different things, prefer `UINT64_C(0)`, `UINT64_C(1)`, `UINT64_C(2)`, etc over `ul`

semihalf-barnas-michal replied:
I've only reformatted this with clang-format, but I will change this. Good to know about these macros, didn't know them.

yperess commented on include/zephyr/drivers/espi_emul.h:
```diff
@@ -1,185 +0,0 @@
-/*
```
Please don't delete the file, we should instead leave an empty stub here with a warning that has:

```
#warning "<zephyr/espi_emul.h> is deprecated, please include <zephyr/espi/espi_emul.h>"
#include <zephyr/espi/espi_emul.h>
```

yperess commented on include/zephyr/drivers/espi/espi.h:
```diff
@@ -101,9 +101,9 @@ enum espi_io_mode {
  */
```
Please don't delete the old file, we should instead leave an empty stub here with a warning that has:

```
#warning "<zephyr/espi.h> is deprecated, please include <zephyr/espi/espi.h>"
#include <zephyr/espi/emul.h>
```

yperess reviewed (changes requested):
Please add documentation to `doc/hardware/peripherals/espi.rst`

barnas-michal committed:
espi: changes types of internal eSPI macros

Instead of using the "ul" suffix, use the UINTx_C() macro to assign values
to the defines.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
espi: add documentation about eSPI emulators

Add some information about availability of the eSPI controller and host
emulators with description of possibilities they give.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

cfriedt commented:
Nothing here jumps out at me as being bad but I'm not terribly involved in espi.

Are @albertofloyd, @VenkatKotakonda, and @jvasanth1 able to review?
